File: Linear Control Course/Chapter7/Lesson2/Chapter7_Lesson2.cpp
#include <iostream>
#include <complex>
#include <Eigen/Dense>

std::string classifyEigenvalues(const Eigen::VectorXcd& evals) {
    bool hasPositive = false;
    bool hasImagAxisRepeated = false;
    int imagAxisCount = 0;

    for (int i = 0; i < evals.size(); ++i) {
        double real_part = evals(i).real();
        double imag_part = evals(i).imag();

        if (real_part > 0.0) {
            hasPositive = true;
        }
        if (std::abs(real_part) < 1e-9 && std::abs(imag_part) > 1e-9) {
            imagAxisCount++;
        }
    }

    if (hasPositive) return "unstable";
    if (imagAxisCount >= 1) {
        // Here we ignore multiplicity check for simplicity
        return "marginally stable or oscillatory";
    }
    return "asymptotically stable";
}

int main() {
    double J = 0.01;
    double b = 0.1;
    double k = 2.0;   // stiffness from PD-like position control

    Eigen::Matrix2d A;
    A << 0.0,              1.0,
          -k / J,   -b / J;

    Eigen::EigenSolver<Eigen::Matrix2d> es(A);
    Eigen::VectorXcd evals = es.eigenvalues();

    std::cout << "Eigenvalues of A:\n";
    for (int i = 0; i < evals.size(); ++i) {
        std::cout << evals(i) << std::endl;
    }

    std::cout << "Stability: " << classifyEigenvalues(evals) << std::endl;

    // In a ROS controller, this logic could be called whenever gains change
    // to ensure the joint-level closed-loop dynamics remain stable.
    return 0;
}

File: Linear Control Course/Chapter7/Lesson2/Chapter7_Lesson2.java
// Example using Apache Commons Math
import org.apache.commons.math3.linear.RealMatrix;
import org.apache.commons.math3.linear.Array2DRowRealMatrix;
import org.apache.commons.math3.linear.EigenDecomposition;

public class StabilityCheck {

    public static String classifyEigenvalues(double[] real, double[] imag) {
        boolean hasPositive = false;
        boolean hasImagAxis = false;

        for (int i = 0; i < real.length; i++) {
            if (real[i] > 0.0) {
                hasPositive = true;
            }
            if (Math.abs(real[i]) < 1e-9 && Math.abs(imag[i]) > 1e-9) {
                hasImagAxis = true;
            }
        }

        if (hasPositive) return "unstable";
        if (hasImagAxis) return "marginally stable (oscillatory)";
        return "asymptotically stable";
    }

    public static void main(String[] args) {
        double J = 0.01;
        double b = 0.1;
        double k = 3.0;

        double[][] dataA = {
                {0.0,          1.0},
                {-k / J,  -b / J}
        };
        RealMatrix A = new Array2DRowRealMatrix(dataA);
        EigenDecomposition ed = new EigenDecomposition(A);

        int n = A.getRowDimension();
        double[] real = new double[n];
        double[] imag = new double[n];

        for (int i = 0; i < n; i++) {
            real[i] = ed.getRealEigenvalue(i);
            imag[i] = ed.getImagEigenvalue(i);
            System.out.println("lambda_" + i + " = " + real[i] + " + j" + imag[i]);
        }

        String stability = classifyEigenvalues(real, imag);
        System.out.println("Stability: " + stability);

        // This pattern can be embedded in Java-based robotic control loops
        // to verify that gain updates preserve closed-loop stability.
    }
}
